Remotion/Data/DomainObjects/Queries/Configuration/QueryDefinition.cs

```diff
@@ -131,10 +131,10 @@ public QueryDefinition (string queryID, StorageProviderDefinition storageProvide
     ArgumentUtility.CheckNotNullOrEmpty("statement", statement);
     ArgumentUtility.CheckValidEnumValue("queryType", queryType);
 
-    if (queryType == QueryType.Scalar && collectionType != null)
+    if (queryType is QueryType.ScalarReadOnly or QueryType.ScalarReadWrite && collectionType != null)
       throw new ArgumentException(string.Format("The scalar query '{0}' must not specify a collectionType.", queryID), "collectionType");
 
-    if (queryType == QueryType.Collection && collectionType == null)
+    if (queryType is QueryType.CollectionReadOnly or QueryType.CollectionReadWrite && collectionType == null)
       collectionType = typeof(DomainObjectCollection);
 
```

Comment:
Are you sure that this is a missing argument checks? Do we need to check whether cusom queries have a collection or not? I would understand custom queries as queries which can do both because they are cutomizable

Reply:
CollectionType must be a DomainObjectCollection which can only hold types of DomainObject. A custom query is short for custom projection query which cannot return types of DomainObject but rather simple dto types. Based on these constraints, QueryType.Custom is not compatible with setting a CollectionType. Custom queries where introduced in a set of second generation changes and some of the older types apparently did not get updated properly. Also, all develper-grade APIs set CollectionType=null for QueryType.Custom, except for the XML-based loading where this constraint was overlooked.

Reply:
Any input?